src/net_processing.cpp

```diff
@@ -2402,15 +2402,20 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
 
         std::map<uint256, std::pair<NodeId, std::list<QueuedBlock>::iterator> >::iterator blockInFlightIt = mapBlocksInFlight.find(pindex->GetBlockHash());
         bool fAlreadyInFlight = blockInFlightIt != mapBlocksInFlight.end();
+        bool in_flight_from_same_peer = fAlreadyInFlight ? blockInFlightIt->second.first == pfrom->GetId() : false;
 
         if (pindex->nStatus & BLOCK_HAVE_DATA) // Nothing to do here
             return true;
 
-        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know something better
+        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know same or better
                 pindex->nTx != 0) { // We had this block at some point, but pruned it
-            if (fAlreadyInFlight) {
-                // We requested this block for some reason, but our mempool will probably be useless
+            if ((fAlreadyInFlight && in_flight_from_same_peer)
+                    || pindex->nChainWork == chainActive.Tip()->nChainWork) {
```

Comment:
I think this will actually request the full block that *is* the known chaintip. Probably need to filter it by blockhash.

Reply:
oh, maybe not. two lines or so above we have this check:

if (pindex->nStatus & BLOCK_HAVE_DATA) // Nothing to do here
            return true;

Reply:
I dont think we want to request the block if its already in flight from another peer, even if it has equal work with our tip.